### PR TITLE
feat(agentic): no-untrusted-content-in-prompt across all three SDKs

### PR DESCRIPTION
**Pull request by ofri-peretz**
## Summary

W3/A3 of `ORM-AGENTIC-RULE-PLAN.md` — the headline CWE-1427 rule of the agentic family, shipped to all three raw inference SDKs from a shared `createSystemPromptInjectionRule` factory.

A system prompt is instruction text. Whatever is spliced into it is read by the model as **instructions**, not as data, so anyone who controls that value controls the agent. Same argument as `mcp-sdk-security/no-tool-description-injection`, one layer down: both are model-facing instruction surfaces that must be static.

Covers both shapes the raw SDKs use:

| Shape | Example |
|---|---|
| Named option | `system` (Anthropic), `instructions` (OpenAI), `systemInstruction` (Gemini, incl. nested `parts[].text`) |
| Message array | `messages: [{ role: 'system', content }]` |

## The taxonomy work is the interesting part

`vercel-ai-security/no-dynamic-system-prompt` already covers this ground for the Vercel AI SDK. Proving these don't collide took two rounds.

**Round 1 — my first cut was wrong.** I gated on the leaf method name `create`, which is shared by every SDK in this family. Measured:

```
line 3  o/no-untrusted-content-in-prompt
line 3  a/no-untrusted-content-in-prompt
COLLISION
```

A file importing both OpenAI and Anthropic reported **one line twice** — exactly what §1.1 forbids.

**Round 2 — gate on the qualified member path.** `messages.create` (Anthropic) vs `completions.create` / `responses.create` (OpenAI) vs `generateContent` (Gemini). A bare `generateText(...)` has no member path at all, which is what keeps the Vercel AI SDK's form with its own plugin.

Re-verified by linting one file that imports **all four** SDKs and uses every shape:

```
line 5  v/no-dynamic-system-prompt
line 6  a/no-untrusted-content-in-prompt
line 7  o/no-untrusted-content-in-prompt
line 8  o/no-untrusted-content-in-prompt
line 9  g/no-untrusted-content-in-prompt
✅ NO LINE REPORTED TWICE
```

## A bare identifier counts as static

`system: SYSTEM_PROMPT` is the correct pattern and by far the most common one; chasing it is the data-flow analysis these rules exist to avoid, and reporting it would make the rule unusable.

But **only as the whole value.** My first version applied the exemption inside concatenations too, so `'You are ' + role` went silent — the exact shape the rule exists to catch. Its own test caught it; `isStaticText` now has no identifier escape hatch.

## `strict` only

Unlike the credential rules in #402/#403, this one has a genuine false-positive shape: a system prompt interpolating today's date is not an injection and the rule cannot tell the difference. Per §1.6 it joins `strict` and waits for the corpus measurement before `recommended`. That required splitting the previously-shared config object in all three plugins, and the preset placement is asserted in each barrel lock so a later refactor can't promote it silently.

## Test plan

- [x] `@interlace/eslint-devkit` — 49 tests on the factory, **100/100/100/100**, no coverage pragmas. Dual-layer: L2 on `isStaticInstruction` / `staticKey` / `isSystemMessage` / `messageContent` / `calleePath`, L1 through a synthetic SDK.
- [x] openai 48, anthropic 45, gemini 46 — each with the bare-function boundary case as an explicit `valid`.
- [x] `turbo run test` 58/58 · `turbo run typecheck` 53/53 · `test:scripts` 1070 · `ecosystem-integrity` 756 · rule-count lock 14.
- [x] `oxlint` exit 0 · `lint:md` 0 issues across 873 files.
- [x] Catalogs regenerated in order: published rule count 462 → 465.

## After merge

`strict` adopters of the three agentic plugins get CWE-1427 coverage. `minimal` and `recommended` are unchanged. `vercel-ai-security` is untouched.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added the `no-untrusted-content-in-prompt` security rule for Anthropic, Gemini, and OpenAI SDK calls.
  * Detects dynamically constructed system prompts and flags potentially untrusted content.
  * Enabled the rule in each plugin’s strict configuration; minimal and recommended configurations are unchanged.
* **Documentation**
  * Added usage guidance and secure coding examples for all three integrations.
* **Tests**
  * Added comprehensive coverage for static, dynamic, and excluded prompt scenarios.
* **Chores**
  * Updated rule manifests and published rule statistics.

### CODE REVIEW

File: packages/eslint-plugin-anthropic-security/docs/rules/no-untrusted-content-in-prompt.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+## Why
+
+A system prompt is instruction text. Whatever is spliced into it is read by the model as *instructions*, not as data — so anyone who controls that value controls the agent's rules. The model has no way to tell "this part of my instructions came from a stranger" once the string is assembled.
+
+This is the same argument as `mcp-sdk-security/no-tool-description-injection`, one layer down: tool descriptions and system prompts are both model-facing instruction surfaces, and both must be static.
+
+## What counts as static
+
+A string literal, a template with no interpolations, or a concatenation of those. **A bare identifier also counts** — `instructions: SYSTEM_PROMPT` is the correct pattern and by far the most common one, and following it is the data-flow analysis this rule deliberately avoids. The interpolation is where the injection is visible in one place, and that is what gets reported.
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply packages/eslint-plugin-anthropic-security/docs/rules/no-untrusted-content-in-prompt.md]: Fixed in a2979fb4 — the doc now says `system: SYSTEM_PROMPT`. This was a real copy-paste from the OpenAI doc: anthropic reads `system`, so anyone following the old text would have written a call the rule stays silent on.

File: packages/eslint-plugin-gemini-security/docs/rules/no-untrusted-content-in-prompt.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+## Why
+
+A system prompt is instruction text. Whatever is spliced into it is read by the model as *instructions*, not as data — so anyone who controls that value controls the agent's rules. The model has no way to tell "this part of my instructions came from a stranger" once the string is assembled.
+
+This is the same argument as `mcp-sdk-security/no-tool-description-injection`, one layer down: tool descriptions and system prompts are both model-facing instruction surfaces, and both must be static.
+
+## What counts as static
+
+A string literal, a template with no interpolations, or a concatenation of those. **A bare identifier also counts** — `instructions: SYSTEM_PROMPT` is the correct pattern and by far the most common one, and following it is the data-flow analysis this rule deliberately avoids. The interpolation is where the injection is visible in one place, and that is what gets reported.
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply packages/eslint-plugin-gemini-security/docs/rules/no-untrusted-content-in-prompt.md]: Fixed in a2979fb4 — now `systemInstruction: SYSTEM_PROMPT`, matching this plugin's `systemPromptProps`.

File: packages/eslint-plugin-openai-security/src/rules/no-untrusted-content-in-prompt/no-untrusted-content-in-prompt.test.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        errors: [{ messageId: 'untrustedSystemPrompt' }],
+      },
+      {
+        name: 'a system prompt built by a call',
+        code: SDK + 'client.responses.create({ instructions: buildPrompt(user) });',
+        errors: [{ messageId: 'untrustedSystemPrompt' }],
+      },
+    ],
+  });
+});
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply packages/eslint-plugin-openai-security/src/rules/no-untrusted-content-in-prompt/no-untrusted-content-in-prompt.test.ts:90]: Added in a2979fb4, both the invalid and the valid case. Verified it is a real lock rather than a passing assertion: dropping `completions.create` from `requestPaths` turns the new case red (`Should have 1 error but had 0`) and leaves the other ten green.

File: packages/eslint-plugin-anthropic-security/src/rules/no-untrusted-content-in-prompt/no-untrusted-content-in-prompt.test.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        errors: [{ messageId: 'untrustedSystemPrompt' }],
+      },
+      {
+        name: 'a system prompt built by a call',
+        code: SDK + 'client.messages.create({ system: buildPrompt(user) });',
+        errors: [{ messageId: 'untrustedSystemPrompt' }],
+      },
+    ],
+  });
+});
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply packages/eslint-plugin-anthropic-security/src/rules/no-untrusted-content-in-prompt/no-untrusted-content-in-prompt.test.ts:81]: Added in a2979fb4. Also added the matching `generateContentStream` case to the Gemini plugin, which had the same second-path gap, plus its nested `parts` shape.

File: packages/eslint-devkit/src/rule-creation/system-prompt-injection-rule.ts [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+   * `messages.create` and OpenAI's `completions.create` are distinct; a bare
+   * `generateText(...)` has no member path at all, which is what keeps
+   * vercel-ai-security's shape out.
+   */
+  requestPaths: readonly string[];
+  /** Option names that hold the system prompt directly, e.g. `system`. */
+  systemPromptProps: readonly string[];
+  docsUrl: string;
+  documentationLink: string;
+}
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply packages/eslint-devkit/src/rule-creation/system-prompt-injection-rule.ts:75]: Added in a2979fb4 — the interface doc states the three-part contract (modules gate the file, requestPaths select the calls, systemPromptProps names the option) and points at the requestPaths note for why the path rather than the leaf method.

File: packages/eslint-plugin-anthropic-security/docs/rules/no-untrusted-content-in-prompt.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+tags: ['security','anthropic']
+category: security
+severity: high
+cwe: CWE-1427
+autofix: false
+---
+
+# no-untrusted-content-in-prompt
+
+> Disallow untrusted content built into a Anthropic system prompt.
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply packages/eslint-plugin-anthropic-security/docs/rules/no-untrusted-content-in-prompt.md]: Fixed in a2979fb4, at the source rather than in the two documents: the shared factory now emits "the <vendor> system prompt", so all three docs and the rule metadata stay consistent and no per-vendor article is needed.

File: apps/docs/src/data/plugin-stats.json [hunk 7]
```diff
@@ -120,33 +120,33 @@
...
-      "published": true
-    },
     {
       "name": "eslint-plugin-anthropic-security",
-      "rules": 2,
+      "rules": 3,
       "description": "ESLint plugin for Anthropic SDK security — catches hardcoded API keys in the Claude client options",
       "category": "security",
       "version": "0.1.1",
       "published": true
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply apps/docs/src/data/plugin-stats.json:129]: Updated in a2979fb4 — the descriptions come from each package.json, so I edited those and regenerated plugin-stats.json. All three now list their full rule set instead of one rule.

File: packages/eslint-devkit/src/rule-creation/system-prompt-injection-rule.ts [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+          const request = node.arguments[0];
+          if (request?.type === AST_NODE_TYPES.ObjectExpression) inspectRequest(request);
+        },
+
+        // Judged at exit so the gate does not depend on statement order.
+        'Program:exit'() {
+          if (!importsSdk) return;
+          for (const node of candidates) {
+            context.report({ node, messageId: 'untrustedSystemPrompt' });
+          }
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply packages/eslint-devkit/src/rule-creation/system-prompt-injection-rule.ts:277]: Correct diagnosis, deliberately not fixed — and now documented instead of implied.

The file-level gate is the pattern the whole SDK-rule family uses (`createSdkApiKeyRule` #402, `createBrowserEscapeHatchRule` #403, both shipped). Resolving the callee root back to the import means following a client through assignment, destructuring, factory returns and re-export — the cross-scope data-flow analysis these rules exist to avoid. The cost is not only complexity: a resolver that cannot see the binding has to choose between reporting anyway (no gain) or going silent (a false negative that looks like a clean scan, which is the worst failure mode we have in this repo).

So the trade is stated in the rule docs under a new **Known limitations** section in a2979fb4, along with the matching false negative on the other side (a prompt assembled before the call). The rule ships in `strict` only; W6 measures the real rate against the corpus, and that measurement — not an estimate — decides whether it earns `recommended` and whether binding is worth building.

One thing I did not do, from the suggestion: I did not add a no-report case for an unrelated local receiver. That test would assert behaviour the rule does not have — it reports there today. Pinning a false expectation is worse than the gap.

File: packages/eslint-plugin-anthropic-security/src/rules/no-untrusted-content-in-prompt/index.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+export const noUntrustedContentInPrompt = createSystemPromptInjectionRule({
+  ruleName: 'no-untrusted-content-in-prompt',
+  vendor: 'Anthropic',
+  modules: ['@anthropic-ai'],
+  requestPaths: ['messages.create', 'messages.stream'],
+  systemPromptProps: ['system'],
+  docsUrl:
+    'https://github.com/ofri-peretz/eslint/blob/main/packages/eslint-plugin-anthropic-security/docs/rules/no-untrusted-content-in-prompt.md',
+  documentationLink: 'https://docs.anthropic.com/en/docs/system-prompts',
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply packages/eslint-plugin-anthropic-security/src/rules/no-untrusted-content-in-prompt/index.ts:30]: Correct diagnosis, deliberately not fixed — and now documented instead of implied.

The file-level gate is the pattern the whole SDK-rule family uses (`createSdkApiKeyRule` #402, `createBrowserEscapeHatchRule` #403, both shipped). Resolving the callee root back to the import means following a client through assignment, destructuring, factory returns and re-export — the cross-scope data-flow analysis these rules exist to avoid. The cost is not only complexity: a resolver that cannot see the binding has to choose between reporting anyway (no gain) or going silent (a false negative that looks like a clean scan, which is the worst failure mode we have in this repo).

So the trade is stated in the rule docs under a new **Known limitations** section in a2979fb4, along with the matching false negative on the other side (a prompt assembled before the call). The rule ships in `strict` only; W6 measures the real rate against the corpus, and that measurement — not an estimate — decides whether it earns `recommended` and whether binding is worth building.

One thing I did not do, from the suggestion: I did not add a no-report case for an unrelated local receiver. That test would assert behaviour the rule does not have — it reports there today. Pinning a false expectation is worse than the gap.
